### PR TITLE
Dimensions: fall back to offsetWidth/Height for border-box in IE

### PR DESCRIPTION
**Pull request by timmywil**
### Summary ###
Fixes gh-4102

If box-sizing is unreliable and box-sizing is border-box, we can use offsetWidth and offsetHeight in most situations. In the situations we can't, box-sizing seems to be reliable and we can use computed CSS. Specifically, in these cases:

- Child of hidden element
- Disconnected element

`valueIsBorderBox` should be `false` for SVG in IE9-11.

The downside here is that we lose fractional values for border-box dimensions in IE9-11. I can live with that, but open to review.

- I first added a commit to include the IE launcher for debugging IE issues on Windows machines, which came in handy for this.
- Also removed a couple unused variables.

### Checklist ###

* [x] All authors have signed the CLA at https://cla.js.foundation/jquery/jquery
* [x] New tests have been added to show the fix or feature works
* [x] Grunt build and unit tests pass locally with these changes

With @gibson042's help, got it down to +9.

### CODE REVIEW

File: src/css.js [hunk 1]
```diff
@@ -133,21 +130,33 @@ function getWidthOrHeight( elem, dimension, extra ) {
...
 	// Support: Android <=4.1 - 4.3 only
 	// Also use offsetWidth/offsetHeight for misreported inline dimensions (gh-3602)
-	if ( val === "auto" ||
-		!parseFloat( val ) && jQuery.css( elem, "display", false, styles ) === "inline" ) {
-
-		val = elem[ "offset" + dimension[ 0 ].toUpperCase() + dimension.slice( 1 ) ];
+	// Support: IE 9-11 only
+	// Also use offsetWidth/offsetHeight for when box sizing is unreliable
+	// Do not use offset for elements that don't have it (e.g. SVG)
+	if ( typeof elem[ offsetProp ] !== "undefined" &&
```
Comment (mgol) [inline src/css.js]: Would moving the `typeof` check to the end of the `if` make sense to avoid computing it if not needed?

File: test/data/testsuite.css [hunk 2]
```diff
@@ -142,3 +142,7 @@ section { background:#f0f; display:block; }
 	padding: 0;
 	margin: 0;
 }
+
+.border-box, .border-box * {
+	box-sizing: border-box;
+}
```
Comment (mgol) [inline test/data/testsuite.css]: That's a total nit but why not:
```css
.border-box {
	box-sizing: border-box;
}
.border-box * {
	box-sizing: inherit;
}
```
? It's easier to override the behavior in sub-trees then.

File: src/css.js [hunk 3]
```diff
@@ -133,21 +130,33 @@ function getWidthOrHeight( elem, dimension, extra ) {
...
+			!parseFloat( val ) && jQuery.css( elem, "display", false, styles ) === "inline" ) ) {
+
+		offsetVal = elem[ offsetProp ];
+
+		// Support: IE 9-11 only
+		// offsetWidth/offsetHeight is zero for disconnect elements
+		// and children of hidden elements
+		if ( offsetVal || !parseFloat( val ) ) {
+			val = offsetVal;
+		}
```
Comment (gibson042) [inline src/css.js]: Big size benefit from dropping `offsetVal` entirely and doing this like
```js
val = elem[ offsetProp ] || parseFloat( val ) || 0;
```
Plus one more from also dropping `offsetProp`:
```js
val = elem[ "offset" + dimension[ 0 ].toUpperCase() + dimension.slice( 1 ) ] ||
	parseFloat( val ) || 0;
```

File: src/css.js (hunk 1 above)
Comment (timmywil) [reply src/css.js]: It's my understanding that `typeof` doesn't cause a compute, but I'll double check that. Could also look at `hasOwnProperty`.

[the remark this answers is not in the document]
Comment (timmywil) [reply src/css.js]: Looked at this again. That's actually the same behavior as before, although it's harder to see. We apparently want it to be false there. I shouldn't have said that the computed value for SVG could be trusted to be border box in IE 9-11.

File: src/css.js [hunk 4]
```diff
@@ -133,21 +130,33 @@ function getWidthOrHeight( elem, dimension, extra ) {
 		val = "auto";
 	}
 
-	// Check for style in case a browser which returns unreliable values
-	// for getComputedStyle silently falls back to the reliable elem.style
-	valueIsBorderBox = valueIsBorderBox &&
-		( support.boxSizingReliable() || val === elem.style[ dimension ] );
+	isBorderBox = jQuery.css( elem, "boxSizing", false, styles ) === "border-box";
+	valueIsBorderBox = isBorderBox && support.boxSizingReliable();
+	offsetProp = "offset" + dimension[ 0 ].toUpperCase() + dimension.slice( 1 );
```
[the remark this answers is not in the document]
Comment (timmywil) [reply src/css.js]: Fair enough

File: src/css.js (hunk 3 above)
Comment (timmywil) [reply src/css.js]: I can do the first, but still need the second to check for existence of the property.

File: test/unit/dimensions.js [hunk 5]
```diff
@@ -646,7 +646,9 @@ QUnit.test( "width/height on a table row with phantom borders (gh-3698)", functi
 	} );
 } );
 
-QUnit.test( "interaction with scrollbars (gh-3589)", function( assert ) {
+// Because we fallback to offsetWidth/offsetHeight,
+// IE9-11 does not report fractional values
+QUnit[ jQuery.support.boxSizingReliable() ? "test" : "skip" ]( "interaction with scrollbars (gh-3589)", function( assert ) {
```
[the remark this answers is not in the document]
Comment (timmywil) [reply test/unit/dimensions.js]: These tests only ensure that fractional values are maintained. Is there another issue your concerned about or do you think fractional values are more important than I do?

File: test/data/testsuite.css (hunk 2 above)
Comment (timmywil) [reply test/data/testsuite.css]: Never occurred to me, but the point of this is to explicitly set `border-box` on all of the elements. Allow them to overridden seems like it defeats the purpose.

**Comment by timmywil** [comment]
> How would you feel about the approach from #4102 (comment) ?

The issue also applies to `textarea`, so limiting to table elements doesn't fix all issues. Also, I prefer a solution that doesn't require checking `nodeName`. Even though some table elements are given special consideration in the spec as offset parents, the solution still feels like it's in the same realm as browser sniffing.

We talked about `|| val === elem.style[ dimension ]` in the meeting. I see it as an enigmatic artifact from old code. It happens to work for a lot of cases, but has no real relation to what we want. What it's doing is checking if the computed value happens to be the same as the inline style and deciding that the value can be treated as border-box because it *probably* doesn't matter, but we've discovered here that's a risky assumption. Patching it up to include the table case (or the case where both values are zero) isn't really solving the problem as there could be more cases to cover that we haven't thought of yet. It's just adding band-aids to an already confusing line of code that doesn't have any logical bearing on whether a value is border-box. I'd much prefer something more explicit. I know that's difficult to accomplish given IE's *loose* relationship with computed border-box, but this is an attempt at addressing the real behavior in IE 9-11 and documenting that behavior with comments.

File: src/css.js (hunk 1 above)
Comment (timmywil) [reply src/css.js]: You're right, it does cause a compute. I switched to `hasOwnProperty`, but I think I'll keep it at the front because I'm concerned that `jQuery.css( elem, "display", false, styles ) === "inline"` might be a little more expensive.

Comment (timmywil) [reply src/css.js]: Sigh, `hasOwnProperty` returned `false` in Chrome despite having the property on inline elements. I just did what you suggested.

File: test/unit/dimensions.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (gibson042) [reply test/unit/dimensions.js]: The tests are not for ensuring preservation of fractional values, but rather for verifying that `.{inner,outer}{Width,Height}` work correctly with scrollable elements as both getters and setters per #3589... doing so in all cases just happens to be tricky because the [fix](https://github.com/jquery/jquery/commit/80f1c8239e9d5f793f3e54e0cb8d7bd0747e4856#diff-414c8f59bff0d1b63680b64763d8c529R134) required arithmetic mixing possibly fractional CSS values with always-integer offsetWidth/offsetHeight, which is why that test block includes fractional CSS values—but note that it is explicitly uses `% 1` to work even on browsers that don't support them.

These tests are complex but IMO necessary, and they're working now. Any change that breaks them makes me uncomfortable, and any change that bypasses them is just begging for trouble.

File: src/css.js (hunk 3 above)
Comment (gibson042) [reply src/css.js]: You can check for property existence by repeating `elem[ "offset" + dimension[ 0 ].toUpperCase() + dimension.slice( 1 ) ]`, which works in our favor here.

**Comment by gibson042** [comment]
> We talked about `|| val === elem.style[ dimension ]` in the meeting. I see it as an enigmatic artifact from old code. It happens to work for a lot of cases, but has no real relation to what we want. What it's doing is checking if the computed value happens to be the same as the inline style and deciding that the value can be treated as border-box because it _probably_ doesn't matter, but we've discovered here that's a risky assumption.

I agree with you here, but I'm trying to gently usher IE out the door, and bringing in such a big hammer—especially one that needs workarounds on workarounds for things like SVG—runs the risk of inviting more strain on a weak point.

> Patching it up to include the table case (or the case where both values are zero) isn't really solving the problem as there could be more cases to cover that we haven't thought of yet. It's just adding band-aids to an already confusing line of code that doesn't have any logical bearing on whether a value is border-box. I'd much prefer something more explicit. I know that's difficult to accomplish given IE's _loose_ relationship with computed border-box, but this is an attempt at addressing the real behavior in IE 9-11 and documenting that behavior with comments.

OK, fair point. I'll re-review with that in mind.

File: test/unit/dimensions.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (gibson042) [reply test/unit/dimensions.js]: Let's run the tests in all environments but force them to be integer-only in IE like so:
```patch
-		fraction = jQuery( "<div style='width:4.5px;'/>" ).appendTo( parent ).width() % 1,
+		fraction = jQuery.support.boxSizingReliable() ?
+			jQuery( "<div style='width:4.5px;'/>" ).appendTo( parent ).width() % 1 :
+			0,
```

File: src/css.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (gibson042) [reply src/css.js]: But SVG elements _do_ respect box-sizing in IE: https://jsbin.com/zafagupixi
Also—and I had forgotten this until just now—`boxSizingUnreliable` really seems to track a specific misbehavior in the form of always reporting content box dimensions, which we can probably use to our advantage here.

But if you really want to use offsetWidth/offsetHeight, what would you say to a variation on something like this?
```js
// Fall back to offsetWidth/offsetHeight when value is "auto"
// This happens for inline elements with no explicit setting (gh-3571)
// Support: IE 9-11 only
// Also use offsetWidth/offsetHeight for when box sizing is unreliable
// Support: Android <=4.1 - 4.3 only
// Also use offsetWidth/offsetHeight for misreported inline dimensions (gh-3602)
if ( isBorderBox !== valueIsBorderBox || val === "auto" ||
		!parseFloat( val ) && jQuery.css( elem, "display", false, styles ) === "inline" ) {

	// Some elements don't have offsetWidth/offsetHeight (e.g., SVG)
	// But where available, it is a means of reading (non-fractional) border box dimensions
	if ( elem[ offsetProp ] > 0 ) {
		val = elem[ offsetProp ];
		valueIsBorderBox = true;
	}
}
```

File: src/css.js (hunk 3 above)
Comment (timmywil) [reply src/css.js]: I don't think we should do that, even if it's smaller.

File: src/css.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (timmywil) [reply src/css.js]: Are you sure you tried that in IE? I see a different result than in Chrome.

File: test/unit/dimensions.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (timmywil) [reply test/unit/dimensions.js]: Thanks for the explanation. I gave up on these tests too quickly then.

File: src/css.js (hunk 3 above)
Comment (timmywil) [reply src/css.js]: Although it may be moot now. Now that it falls back to val, `undefined` is fine.

Comment (timmywil) [reply src/css.js]: Oh wait, that's only if we want `valueIsBorderBox` to be `true` for SVG. From my testing, it seems like it should be `false`. I'll wait for your reply on the jsbin.

File: src/css.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (timmywil) [reply src/css.js]: Or maybe I'm confusing your point. Are we saying the same thing in different ways? The render seems fine, it's just that computed value for border-box is content-box for SVG. Your code suggestion should work.

File: test/unit/dimensions.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (timmywil) [reply test/unit/dimensions.js]: Your suggestion works.

File: src/css.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (timmywil) [reply src/css.js]: Actually, I just tried that and it broke the inline/disconnected tests. We want `valueIsBorderBox` to still be set to true for inline and disconnected elements, but false for SVG.

Comment (timmywil) [reply src/css.js]: @mgol brilliant

Comment (timmywil) [reply src/css.js]: I like it. It brings it down to +6 as well.

File: test/unit/dimensions.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (timmywil) [reply test/unit/dimensions.js]: Fixed IE9 by accounting for the element shrinking at the test level rather than trying to adjust the width. Our code compensates for scrollbars + border-box + position: absolute in IE9 only and it's like it was trying to compensate twice.

File: src/css.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (gibson042) [reply src/css.js]: I don't think a statement like "we want `valueIsBorderBox` to still be set to true for inline and disconnected elements, but false for SVG" is accurate. What we want is proper input for `boxModelAdjustment`, which returns the amount by which the width or height of a specified box (conveyed by `box`) exceeds a CSS computed width or height (subject to a box model indicated by `isBorderBox`). `boxModelAdjustment` could probably use some clarifying renames and/or comments, but the intended returned values are as follows:

`box` ＼ `isBorderBox` | false (`content-box`) | true (`border-box`)
---- | ---- | ----
content | 0 | -border - padding
padding | padding (+ scroll gutter) | -border
border | padding + border (+ scroll gutter) | 0
margin | padding + border + margin (+ scroll gutter) | margin

The [tests that failed](https://github.com/jquery/jquery/blob/6153eb0fd401cda90bf2007335cd4338093d38f0/test/unit/dimensions.js#L415) are... bad. They were only passing before this PR because of the inline style check that you're removing, and they're only passing in the current state of this PR because it includes an implicit border-box bias when _attempting_ to use offsetWidth/offsetHeight, even if that attempt fails.

We make no promises about accuracy in those disconnected/hidden cases, but if we want to continue doing our best then the right way would be to trust the computed box-sizing value for disconnected elements.

File: test/unit/dimensions.js [hunk 6]
```diff
@@ -694,13 +696,6 @@ QUnit.test( "interaction with scrollbars (gh-3589)", function( assert ) {
 			borderBox.clone().css( { overflow: "auto" } ).appendTo( parent )[ 0 ].offsetWidth -
 			borderBox[ 0 ].offsetWidth;
 
-	if ( borderBoxLoss > 0 ) {
-		borderBox.css( {
-			width: ( size + borderBoxLoss ) + "px",
-			height: ( size + borderBoxLoss ) + "px"
-		} );
-	}
```
Comment (gibson042) [inline test/unit/dimensions.js]: It took some self-convincing, but this change seems to be OK. However, I do like the original style better because `borderBox.innerWidth(99); … assert.equal(borderBox.innerWidth(), 99)` is more direct than `borderBox.innerWidth(82); … assert.equal(borderBox.innerWidth() + 17, 99)`, especially in comparison with the corresponding `contentBox` assertions (whose values are always 6 less, to account for 1px border width and 2px of padding on each edge).
<details><summary><code>borderBox</code> evaluation diff</summary>

```diff
--- before
+++ after
 size = 100
 borderBoxLoss = 17
-borderBox.css( { width: "117px", height: "117px" } )
 i = 0
 assert.equal( contentBox.innerWidth/innerHeight(), 104 )
 assert.equal( contentBox.outerWidth/outerHeight(), 106 )
-assert.equal( borderBox.innerWidth/innerHeight(), 98 )
-assert.equal( borderBox.outerWidth/outerHeight(), 100 )
+assert.equal( borderBox.innerWidth/innerHeight() + 17, 98 )
+assert.equal( borderBox.outerWidth/outerHeight() + 17, 100 )
 i = 1
 size = 101
-borderBox.innerWidth( 98 + 1 ).innerHeight( 98 + 1 )
+borderBox.innerWidth( 81 + 1 ).innerHeight( 81 + 1 )
 assert.equal( contentBox.innerWidth/innerHeight(), 105 )
 assert.equal( contentBox.outerWidth/outerHeight(), 107 )
-assert.equal( borderBox.innerWidth/innerHeight(), 99 )
-assert.equal( borderBox.outerWidth/outerHeight(), 101 )
+assert.equal( borderBox.innerWidth/innerHeight() + 17, 99 )
+assert.equal( borderBox.outerWidth/outerHeight() + 17, 101 )
 i = 2
 size = 103
-borderBox.outerWidth( 101 + 2 ).outerHeight( 101 + 2 )
+borderBox.outerWidth( 84 + 2 ).outerHeight( 84 + 2 )
 assert.equal( contentBox.innerWidth/innerHeight(), 107 )
 assert.equal( contentBox.outerWidth/outerHeight(), 109 )
-assert.equal( borderBox.innerWidth/innerHeight(), 101 )
-assert.equal( borderBox.outerWidth/outerHeight(), 103 )
+assert.equal( borderBox.innerWidth/innerHeight() + 17, 101 )
+assert.equal( borderBox.outerWidth/outerHeight() + 17, 103 )
```
</details>

Comment (gibson042) [reply test/unit/dimensions.js]: Update: No, these changes are in fact wrong. And we need that initial `.css(…)` call to reveal the bug, because the difference is a constant offset after that.

File: src/css.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (gibson042) [reply src/css.js]: @timmywil this should work, and at a reasonable +9 bytes over master:
```js
var styles = getStyles( elem ),
	val = curCSS( elem, dimension, styles ),
	isBorderBox = jQuery.css( elem, "boxSizing", false, styles ) === "border-box",
	valueIsBorderBox = isBorderBox,
	offsetProp = "offset" + dimension[ 0 ].toUpperCase() + dimension.slice( 1 );
…
// Fall back to offsetWidth/offsetHeight when value is "auto"
// This happens for inline elements with no explicit setting (gh-3571)
// Support: IE 9-11 only
// Also use offsetWidth/offsetHeight for when box-sizing is unreliable
// Support: Android <=4.1 - 4.3 only
// Also use offsetWidth/offsetHeight for misreported inline dimensions (gh-3602)
if ( isBorderBox && !support.boxSizingReliable() || val === "auto" ||
	!parseFloat( val ) && jQuery.css( elem, "display", false, styles ) === "inline" ) {

	if ( elem.getClientRects().length ) {

		// Where available, offsetWidth/offsetHeight approximate border box dimension.
		// Where not available (e.g., SVG), assume unreliable box-sizing and interpret the
		// retrieved value as a content box dimension.
		valueIsBorderBox = offsetProp in elem;
		if ( valueIsBorderBox ) {
			val = elem[ offsetProp ];
		}
	}
}
```
(note the initial assignment of `valueIsBorderBox` to match `isBorderBox`, which will always be preserved for hidden/disconnected elements but will be overridden for visible elements when box-sizing makes gCS dimensions unreliable—matching IE9-11 behavior)

File: test/unit/dimensions.js (hunk 6 above)
Comment (timmywil) [reply test/unit/dimensions.js]: Sorry what bug does it reveal exactly? The element is still being shrunk, but we're making it shrink to the the value we expect in the assertion instead of changing the assertion. I don't really see a logical difference. Either way, we're accounting for the size difference in IE9. I don't really mind your code suggestion. I'm just not seeing any advantage other than a cosmetic one.

File: src/css.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (timmywil) [reply src/css.js]: My statement is still accurate tho, right? The point was simply that the computed value can be trusted for hidden/disconnected but not for SVG, which is maintained in your suggestion. But this does avoid accessing offset for hidden/disconnected when unnecessary so I like it better.

File: test/unit/dimensions.js (hunk 6 above)
Comment (timmywil) [reply test/unit/dimensions.js]: I went ahead and changed it, but I'd still like to understand how the previous was wrong.

File: src/css.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (timmywil) [reply src/css.js]: Updated. I just moved the rects check to the outer if statement as it was the same size.

**Comment by timmywil** [comment]
It's amazing how dense this code is.

File: test/unit/dimensions.js (hunk 6 above)
Comment (gibson042) [reply test/unit/dimensions.js]: I think it's because of the particularly strange manifestation of #3699... we want to have a dimension set before the assertions. Things might still work otherwise, but I definitely saw some passes that should have been failures when I was experimenting. And if I remember correctly from the original ticket, there were scenarios where the reported values didn't agree with the painted dimensions.

Bottom line: "wrong" is probably too strong, but it is more reliable to ensure a set before testing against expectations.

**Comment by timmywil** [comment]
I should be able to get to it today.

**Comment by mgol** [comment]
The test added in this PR still fails in Android Browser 4.0-4.3: http://swarm.jquery.org/result/3143623.

**Comment by timmywil** [comment]
come on android
